server/update-workflow-interface/update-workflow-interface.md

```diff
@@ -0,0 +1,217 @@
+
+### Invalid Update
+
+UpdateRequests can be invalid given the state of the receiving workflow. In such
+cases the UpdateResponse will contain an error indication and payload. Invalid
+updates _must not_ be recorded in workflow history to prevent a malicious issuer
+from causing a denial of service on the workflow by issuing repeated invalid
+UpdateRequests.
+
+### "Failing" Update
+
+UpdateRequests can be semantically acceptable but still fail due to external
+dependencies. In such cases the UpdateResponse will contain an error indication
+and payload.
+
+### Unexpected Disconnect
+
+While the issuer of an update may wish to block on an UpdateResponse, network
+conditions or software failures may occur which forcibly disconnects the update
+issuer from the server at the time that the Update completes _and_ before the
+issuer becomes aware of the Update ID. The issuer must be able to re-issue the
+same UpdateRequest to access the server-generated Update ID.
+
+## gRPC API
+
+The gRPC API comprises two functions, `UpdateWorkflow` and
+`AwaitUpdateWorkflowResponse` to be added to the `WorkflowService` service
+definition. The former is used to invoke an update and follows the standard
+patterns for identifying a workflow execution and carrying arbitrary metadata
+(Header) and payloads (Payloads). The request type for this function includes a
+`wait_timeout` property indicating how long the client is willing to wait for a
+synchronous response once the request has been accepted by the system. This
+value can be set to zero indicating no such waiting should occur.  The
+`request_id` property uniquely identifies the logical request (_not_ the
+issuance of the request) and is intended for use as an idempotency token.
+
+The second function, `AwaitUpdateWorkflowResponse` is used when a client has
+already issued an `UpdateWorkflow` request and has the resulting `update_id`.
+That `update_id` can be provided along with a `wait_timeout` (same semantics as
+above) to re-attach to an in-flight or completed Update and obtain the
+associated UpdateResult in the form of an `UpdateWorkflowResponse`.
+
+
+```protobuf
+service WorkflowService {
+    rpc UpdateWorkflow(UpdateWorkflowRequest) returns (UpdateWorkflowResponse){}
+    rpc AwaitUpdateWorkflowResponse(AwaitUpdateWorkflowResponseRequest) returns (UpdateWorkflowResponse){}
+}
+
+message UpdateWorkflowRequest {
+    temporal.api.common.v1.Header header = 1;
+    string request_id = 2;
+    string namespace = 3;
+    temporal.api.common.v1.WorkflowExecution execution = 4;
+    string update_name = 5;
+    temporal.api.common.v1.Payloads input = 6;
+    google.protobuf.Duration wait_timeout = 7;
+}
+
+message UpdateWorkflowResponse {
+    UpdateResultType result_type = 1;
+    string update_id = 2;
+    temporal.api.common.v1.Payloads output = 3;
+}
+
+enum UpdateResultType {
+    UPDATE_RESULT_TYPE_UNSPECIFIED = 0;
+    UPDATE_RESULT_TYPE_SUCCESS = 1;
+    UPDATE_RESULT_TYPE_ERROR = 2;
```

Comment:
Is this only for workflow-returned errors or does it also represent update timeout, workflow not present, update not handled by workflow, etc?

Reply:
I have a slight preference not to use presence/absence in proto3 just because _everything_ is optional. It's like protocol design with one hand tied behind your back. That said, doing `oneof { Payloads | Failure | Empty }` or something like that is a cleaner protobuf encapsulation (the mapping of oneof into language constructs though ... I'm not sure I want to inflict that on anyone).

Reply:
Ended up with nullable UpdateSuccess and UpdateFailure types/fields with a discriminator enum. While the enum is redundant it seems consistent with existing practices to have it.